server/src/main/java/org/opensearch/indices/IndicesRequestCache.java

```diff
@@ -116,22 +122,35 @@ public final class IndicesRequestCache implements RemovalListener<IndicesRequest
     private final Set<CleanupKey> keysToClean = ConcurrentCollections.newConcurrentSet();
     private final ByteSizeValue size;
     private final TimeValue expire;
-    private final Cache<Key, BytesReference> cache;
+    private final ICache<Key, BytesReference> cache;
     private final Function<ShardId, Optional<CacheEntity>> cacheEntityLookup;
 
-    IndicesRequestCache(Settings settings, Function<ShardId, Optional<CacheEntity>> cacheEntityFunction) {
+    IndicesRequestCache(Settings settings, Function<ShardId, Optional<CacheEntity>> cacheEntityFunction, CacheService cacheService) {
         this.size = INDICES_CACHE_QUERY_SIZE.get(settings);
         this.expire = INDICES_CACHE_QUERY_EXPIRE.exists(settings) ? INDICES_CACHE_QUERY_EXPIRE.get(settings) : null;
         long sizeInBytes = size.getBytes();
-        CacheBuilder<Key, BytesReference> cacheBuilder = CacheBuilder.<Key, BytesReference>builder()
-            .setMaximumWeight(sizeInBytes)
-            .weigher((k, v) -> k.ramBytesUsed() + v.ramBytesUsed())
-            .removalListener(this);
-        if (expire != null) {
-            cacheBuilder.setExpireAfterAccess(expire);
-        }
-        cache = cacheBuilder.build();
+        ToLongBiFunction<Key, BytesReference> weigher = (k, v) -> k.ramBytesUsed() + v.ramBytesUsed();
         this.cacheEntityLookup = cacheEntityFunction;
+        if (FeatureFlags.TIERED_CACHING_SETTING.get(settings)) {
+            this.cache = cacheService.createCache(
+                new CacheConfig.Builder<Key, BytesReference>().setSettings(settings)
+                    .setWeigher((k, v) -> k.ramBytesUsed() + v.ramBytesUsed())
+                    .setValueType(BytesReference.class)
+                    .setKeyType(Key.class)
+                    .setRemovalListener(this)
+                    .build(),
+                CacheType.INDICES_REQUEST_CACHE
+            );
+        } else {
+            ICacheBuilder<Key, BytesReference> builder = new OpenSearchOnHeapCache.Builder<Key, BytesReference>().setSettings(settings)
+                .setMaximumWeightInBytes(sizeInBytes)
+                .setWeigher(weigher)
+                .setRemovalListener(this);
+            if (expire != null) {
+                builder.setExpireAfterAccess(expire);
+            }
+            this.cache = builder.build();
+        }
```

Comment:
To reduce the gap between the feature flagged execution and vanilla, would it make sense to use the `cacheService` to create the on-heap cache when the feature flag is not set as well? 

Would it make sense to move the feature flag check into `CacheService`?

Reply:
On the other hand, once we remove the feature flag, we'll still need to support the old settings for backward compatibility, right? So, I think `CacheService` still needs to understand both the old and new settings anyway.

Reply:
I evaluated this option again.

I had below doubts for which I am looking to get more clarity/suggestions. Will raise revision accordingly.


#### 1. How would a user use the new setting once we deprecate this feature flag setting?
#### Approach 1a
I think one way can that a user sets below cache store name explicitly, we can then consider that he/she is looking to use  the new setting.
 ```
 indices.request.cache.store.name: "on_heap" // user intends to use pluggable caches
// Rest associated onheap setting
indices.request.cache.on_heap.max_size: 100kb
```
  
Otherwise we assume, that user intent is to use older settings (until unless we deprecate this altogether at some point)

 ```
indices.request.cache.store.name: "" // not set which can be the default value.
indices.request.cache.size: 100kb // Old setting is used.
 ```
#### Approach 1b
We can also convert this feature flag into an actual setting and user can set it to true if they want to use pluggable caches or not. This seems a better way to me. So we don't need to remove this condition as such.


#### 2. Also should cacheService start maintaining stats for vanilla cache in case feature flag is turned off? 
I am guessing yes, right?

Reply:
@msfroh I have made the changes as per approach 1a.

Also CacheService will maintain vanilla cache with these changes.